File: python/cadreen/resources/documents.py
from __future__ import annotations

from ..client import HttpClient
from ..types import (
    ListDocumentsResponse,
    Document,
    Pagination,
)


class DocumentsResource:
    def __init__(self, client: HttpClient) -> None:
        self._client = client

    async def list(self) -> ListDocumentsResponse:
        raw = await self._client.get("/api/v1/cadreen/documents")
        documents = [
            Document(
                id=d["id"],
                name=d["name"],
                content_type=d.get("content_type"),
                size=d.get("size"),
                status=d.get("status"),
                created_at=d.get("created_at"),
            )
            for d in raw.get("documents", [])
        ]
        pagination = None
        if raw.get("pagination"):
            p = raw["pagination"]
            pagination = Pagination(limit=p["limit"], offset=p["offset"], has_more=p["has_more"])
        return ListDocumentsResponse(
            documents=documents,
            count=raw.get("count", 0),
            pagination=pagination,
        )

    async def get(self, id: str) -> Document:
        raw = await self._client.get(f"/api/v1/cadreen/documents/{id}")
        return Document(
            id=raw["id"],
            name=raw["name"],
            content_type=raw.get("content_type"),
            size=raw.get("size"),
            status=raw.get("status"),
            created_at=raw.get("created_at"),
        )

File: go/cmd/cadreen/commands/documents.go
package commands

import (
	"encoding/json"
	"fmt"

	"github.com/timothy-billingrails/cadreen-sdks/go/cmd/cadreen/output"

	cadreen "github.com/timothy-billingrails/cadreen-sdks/go/cadreen"

	"github.com/spf13/cobra"
)

var documentsCmd = &cobra.Command{
	Use:   "documents",
	Short: "Manage documents (list, get)",
	Long:  `Manage uploaded documents.`,
}

var documentsListCmd = &cobra.Command{
	Use:   "list",
	Short: "List documents",
	RunE: func(cmd *cobra.Command, args []string) error {
		requireAuth()

		resp, err := rawDo("GET", "/api/v1/cadreen/documents", nil)
		if err != nil {
			return handleAPIError(err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != 200 {
			return handleHTTPError(resp)
		}

		var docs cadreen.ListDocumentsResponse
		if err := json.NewDecoder(resp.Body).Decode(&docs); err != nil {
			return err
		}

		format := outputFormat()
		switch format {
		case output.FormatJSON:
			output.Print(docs, format)
		case output.FormatVerbose:
			printDocumentsVerbose(docs.Documents)
		default:
			printDocumentsText(docs.Documents)
		}

		return nil
	},
}

var documentsGetCmd = &cobra.Command{
	Use:   "get [id]",
	Short: "Get a document by ID",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		requireAuth()

		resp, err := rawDo("GET", "/api/v1/cadreen/documents/"+args[0], nil)
		if err != nil {
			return handleAPIError(err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != 200 {
			return handleHTTPError(resp)
		}

		var doc cadreen.Document
		if err := json.NewDecoder(resp.Body).Decode(&doc); err != nil {
			return err
		}

		format := outputFormat()
		switch format {
		case output.FormatJSON:
			output.Print(doc, format)
		default:
			fmt.Printf("ID:      %s\n", doc.ID)
			fmt.Printf("Name:    %s\n", doc.Name)
			if doc.ContentType != "" {
				fmt.Printf("Type:    %s\n", doc.ContentType)
			}
			if doc.Size > 0 {
				fmt.Printf("Size:    %d bytes\n", doc.Size)
			}
			if doc.Status != "" {
				fmt.Printf("Status:  %s\n", doc.Status)
			}
		}

		return nil
	},
}

func init() {
	documentsCmd.AddCommand(documentsListCmd)
	documentsCmd.AddCommand(documentsGetCmd)
	rootCmd.AddCommand(documentsCmd)
}

func printDocumentsText(docs []cadreen.Document) {
	if len(docs) == 0 {
		fmt.Println("No documents found.")
		return
	}

	fmt.Printf("Documents (%d):\n\n", len(docs))
	for _, d := range docs {
		fmt.Printf("  %s\n", d.Name)
		if d.ContentType != "" {
			fmt.Printf("    Type: %s\n", d.ContentType)
		}
		fmt.Println()
	}
}

func printDocumentsVerbose(docs []cadreen.Document) {
	for _, d := range docs {
		fmt.Printf("ID:      %s\n", d.ID)
		fmt.Printf("Name:    %s\n", d.Name)
		fmt.Printf("Type:    %s\n", d.ContentType)
		fmt.Printf("Size:    %d\n", d.Size)
		fmt.Printf("Status:  %s\n", d.Status)
		fmt.Println()
	}
}
